File: discovr.py
#!/usr/bin/env python3
# discovr.py
import argparse, asyncio, ipaddress, socket, csv, time, sys
from pathlib import Path
from datetime import datetime

DEFAULT_PORTS = [22, 80, 135, 443, 445, 3389]

def guess_os(open_ports):
    p = set(open_ports)
    if 135 in p or 445 in p or 3389 in p: return "Windows (heuristic)"
    if 22 in p: return "Linux/Unix-like (heuristic)"
    if 80 in p or 443 in p: return "Unknown (web)"
    return "Unknown"

async def tcp_probe(host, port, timeout=1.5):
    try:
        await asyncio.wait_for(asyncio.open_connection(host, port), timeout)
        return True
    except:
        return False

async def scan_host(host, ports, sem, timeout):
    open_ports = []
    async with sem:
        for port in ports:
            if await tcp_probe(host, port, timeout):
                open_ports.append(port)
    if not open_ports:
        return None
    return {
        "ip": host,
        "hostname": safe_gethost(host),
        "open_ports": ",".join(map(str, open_ports)),
        "os_guess": guess_os(open_ports),
    }

def safe_gethost(ip):
    try:
        return socket.gethostbyaddr(ip)[0]
    except:
        return ""

async def run_scan(cidr, ports, concurrency, timeout, max_hosts=None, deadline=None):
    net = ipaddress.ip_network(cidr, strict=False)
    targets = [str(ip) for ip in net.hosts()]
    if max_hosts: targets = targets[:max_hosts]
    sem = asyncio.Semaphore(concurrency)
    results = []
    start = time.time()
    for ip in targets:
        if deadline and time.time() > deadline:
            break
        res = await scan_host(ip, ports, sem, timeout)
        if res: results.append(res)
    elapsed = time.time() - start
    return results, elapsed, len(targets)

def write_csv(rows, out_csv):
    out_csv.parent.mkdir(parents=True, exist_ok=True)
    with out_csv.open("w", newline="", encoding="utf-8") as f:
        w = csv.DictWriter(f, fieldnames=["ip","hostname","open_ports","os_guess"])
        w.writeheader()
        for r in rows:
            w.writerow(r)

def write_html(rows, out_html):
    out_html.parent.mkdir(parents=True, exist_ok=True)
    head = """<!doctype html><meta charset="utf-8">
<title>Discovr Report</title>
<style>body{font:14px system-ui, sans-serif;padding:16px}
table{border-collapse:collapse;width:100%}th,td{border:1px solid #ddd;padding:8px}
th{background:#f5f5f5;cursor:pointer}#q{margin:8px 0;padding:8px;width:100%}</style>
<h2>Discovr Report</h2><input id="q" placeholder="Filter...">
<table id="t"><thead><tr><th>IP</th><th>Hostname</th><th>Open Ports</th><th>OS Guess</th></tr></thead><tbody>"""
    body = "".join(
        f"<tr><td>{r['ip']}</td><td>{r['hostname']}</td><td>{r['open_ports']}</td><td>{r['os_guess']}</td></tr>"
        for r in rows
    )
    foot = """</tbody></table>
<script>
const q=document.getElementById('q');const rows=[...document.querySelectorAll('#t tbody tr')];
q.addEventListener('input',()=>{const s=q.value.toLowerCase();rows.forEach(r=>{r.style.display=
r.textContent.toLowerCase().includes(s)?'':'none';});});
</script>"""
    out_html.write_text(head+body+foot, encoding="utf-8")

def main():
    ap = argparse.ArgumentParser(description="Discovr - lightweight asset discovery (CSV only)")
    ap.add_argument("--target", required=True, help="CIDR, e.g. 192.168.1.0/24")
    ap.add_argument("--quick-scan", action="store_true", help="Use quick preset")
    ap.add_argument("--ports", default=",".join(map(str, DEFAULT_PORTS)))
    ap.add_argument("--concurrency", type=int, default=256)
    ap.add_argument("--timeout", type=float, default=1.5)
    ap.add_argument("--max-hosts", type=int, default=None)
    ap.add_argument("--deadline-sec", type=int, default=150, help="Stop around this many seconds")
    ap.add_argument("--out", default="out/report.csv")
    args = ap.parse_args()

    if args.quick_scan:
        args.concurrency = 256
        args.timeout = 1.5
        if not args.max_hosts: args.max_hosts = None  # allow full /24 by default

    ports = [int(p) for p in args.ports.split(",") if p.strip()]
    out_csv = Path(args.out)
    out_html = out_csv.with_suffix(".html")

    print(f"[*] Discovr v0.1 - target={args.target} ports={ports} conc={args.concurrency} timeout={args.timeout}")
    print(f"[*] Starting {datetime.now().isoformat(timespec='seconds')}")

    deadline = time.time() + args.deadline_sec if args.deadline_sec else None
    rows, elapsed, total = asyncio.run(run_scan(args.target, ports, args.concurrency, args.timeout, args.max_hosts, deadline))

    write_csv(rows, out_csv)
    write_html(rows, out_html)

    print(f"[+] Hosts scanned (attempted): {total}")
    print(f"[+] Discovered: {len(rows)}")
    print(f"[*] Exported: {out_csv} and {out_html}")
    print(f"[*] Done in {elapsed:0.2f}s")

if __name__ == "__main__":
    main()

File: discovr.go
package main

import (
	"encoding/csv"
	"fmt"
	"flag"
	"net"
	"os"
	"strings"
	"time"
	"sync"
	"log"
	"path/filepath"
	"html/template"
)

// Define the default ports to scan
var DEFAULT_PORTS = []int{22, 80, 135, 443, 445, 3389}

// Struct for command-line arguments
type Config struct {
	Target     string
	QuickScan  bool
	OutputFile string
	HTMLFile   string
}

// Guess the OS based on open ports
func guessOS(openPorts []int) string {
	p := make(map[int]bool)
	for _, port := range openPorts {
		p[port] = true
	}
	if p[135] || p[445] || p[3389] {
		return "Windows (heuristic)"
	}
	if p[22] {
		return "Linux/Unix-like (heuristic)"
	}
	if p[80] || p[443] {
		return "Unknown (web)"
	}
	return "Unknown"
}

// Probe a TCP connection on a given host and port
func tcpProbe(host string, port int, timeout time.Duration) bool {
	conn, err := net.DialTimeout("tcp", fmt.Sprintf("%s:%d", host, port), timeout)
	if err != nil {
		return false
	}
	conn.Close()
	return true
}

// Scan a single host for open ports
func scanHost(host string, ports []int, timeout time.Duration) (map[string]interface{}, error) {
	var openPorts []int
	for _, port := range ports {
		if tcpProbe(host, port, timeout) {
			openPorts = append(openPorts, port)
		}
	}
	if len(openPorts) == 0 {
		return nil, nil
	}
	hostName, _ := net.LookupAddr(host)
	return map[string]interface{}{
		"ip":         host,
		"hostname":   hostName,
		"open_ports": strings.Join(intToStr(openPorts), ","),
		"os_guess":   guessOS(openPorts),
	}, nil
}

// Convert int array to string array
func intToStr(ints []int) []string {
	strs := make([]string, len(ints))
	for i, v := range ints {
		strs[i] = fmt.Sprintf("%d", v)
	}
	return strs
}

// Generate a list of IP addresses from a CIDR range
func cidrRange(cidr string) []string {
	_, ipnet, err := net.ParseCIDR(cidr)
	if err != nil {
		log.Fatal("Invalid CIDR format:", err)
	}

	var ipAddresses []string
	for ip := ipnet.IP.Mask(ipnet.Mask); ipnet.Contains(ip); increment(ip) {
		ipAddresses = append(ipAddresses, ip.String())
	}
	return ipAddresses
}

// Increment an IP address
func increment(ip net.IP) {
	for j := len(ip) - 1; j >= 0; j-- {
		ip[j]++
		if ip[j] != 0 {
			break
		}
	}
}

// Run the scan on the given CIDR range
func runScan(cidr string, ports []int, concurrency int, timeout time.Duration) ([]map[string]interface{}, time.Duration) {
	// Generate IPs from the CIDR range
	var results []map[string]interface{}
	var targets []string
	for _, ip := range cidrRange(cidr) {
		targets = append(targets, ip)
	}

	var wg sync.WaitGroup
	sem := make(chan struct{}, concurrency)
	start := time.Now()

	for _, target := range targets {
		wg.Add(1)
		go func(ip string) {
			defer wg.Done()
			sem <- struct{}{}
			res, _ := scanHost(ip, ports, timeout)
			if res != nil {
				results = append(results, res)
			}
			<-sem
		}(target)
	}

	wg.Wait()
	elapsed := time.Since(start)

	return results, elapsed
}

// Write the scan results to a CSV file
func writeCSV(rows []map[string]interface{}, outCSV string) {
	// Ensure the directory exists or create it
	dir := filepath.Dir(outCSV)
	if err := os.MkdirAll(dir, 0755); err != nil {
		fmt.Println("Error creating directory:", err)
		return
	}

	// Create the CSV file
	file, err := os.Create(outCSV)
	if err != nil {
		fmt.Println("Error creating CSV file:", err)
		return
	}
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	// Write the headers
	writer.Write([]string{"IP", "Hostname", "Open Ports", "OS Guess"})
	// Write the data rows
	for _, row := range rows {
		writer.Write([]string{
			row["ip"].(string),
			row["hostname"].(string),
			row["open_ports"].(string),
			row["os_guess"].(string),
		})
	}
}

// Write the scan results to an HTML file with search functionality
func writeHTML(rows []map[string]interface{}, outHTML string) {
	// Ensure the directory exists or create it
	dir := filepath.Dir(outHTML)
	if err := os.MkdirAll(dir, 0755); err != nil {
		fmt.Println("Error creating directory:", err)
		return
	}

	// Define the HTML template with a search/filter feature
	tmpl, err := template.New("report").Parse(`
	<!doctype html>
	<html>
	<head>
		<title>Discovr Scan Report</title>
		<style>
			body { font-family: Arial, sans-serif; }
			table { width: 100%; border-collapse: collapse; margin-top: 20px; }
			th, td { padding: 8px; text-align: left; border: 1px solid #ddd; }
			th { background-color: #f2f2f2; }
			#q { margin: 8px 0; padding: 8px; width: 100%; }
		</style>
	</head>
	<body>
		<h1>Discovr Scan Report</h1>
		<input id="q" placeholder="Filter results..." />
		<table>
			<tr><th>IP</th><th>Hostname</th><th>Open Ports</th><th>OS Guess</th></tr>
			{{range .}}
				<tr>
					<td>{{.ip}}</td>
					<td>{{.hostname}}</td>
					<td>{{.open_ports}}</td>
					<td>{{.os_guess}}</td>
				</tr>
			{{end}}
		</table>
		<script>
			const q = document.getElementById('q');
			const rows = [...document.querySelectorAll('table tbody tr')];
			q.addEventListener('input', () => {
				const s = q.value.toLowerCase();
				rows.forEach(row => {
					row.style.display = row.textContent.toLowerCase().includes(s) ? '' : 'none';
				});
			});
		</script>
	</body>
	</html>`)

	if err != nil {
		fmt.Println("Error parsing HTML template:", err)
		return
	}

	// Create the HTML file
	file, err := os.Create(outHTML)
	if err != nil {
		fmt.Println("Error creating HTML file:", err)
		return
	}
	defer file.Close()

	// Execute the template and write to the file
	err = tmpl.Execute(file, rows)
	if err != nil {
		fmt.Println("Error executing HTML template:", err)
	}
}


// Parse command-line arguments
func parseFlags() Config {
	var cfg Config
	flag.StringVar(&cfg.Target, "target", "192.168.1.0/24", "Target IP range (CIDR notation)")
	flag.BoolVar(&cfg.QuickScan, "quick-scan", false, "Enable quick scan (fewer ports)")
	flag.StringVar(&cfg.OutputFile, "out", "outgo/report.csv", "Output CSV file")
	flag.StringVar(&cfg.HTMLFile, "html", "outgo/report.html", "Output HTML file")
	flag.Parse()
	return cfg
}

func main() {
	// Parse command-line arguments
	cfg := parseFlags()

	// Default ports to scan
	var ports []int
	if cfg.QuickScan {
		ports = []int{22, 80, 443} // Limited set of ports for quick scan
	} else {
		ports = DEFAULT_PORTS // Full port set
	}

	// Set a timeout for TCP connections
	timeout := 1 * time.Second

	// Run the scan
	results, elapsed := runScan(cfg.Target, ports, 256, timeout)

	// Write results to CSV
	writeCSV(results, cfg.OutputFile)

	// Write results to HTML
	writeHTML(results, cfg.HTMLFile)

	// Print the scan summary
	fmt.Printf("[+] Hosts scanned: %d\n", len(results))
	fmt.Printf("[+] Discovered: %d\n", len(results))
	fmt.Printf("[+] Exported to: %s and %s\n", cfg.OutputFile, cfg.HTMLFile)
	fmt.Printf("[+] Done in %.2fs\n", elapsed.Seconds())
}
